File: packages/devices-capacitor/ios/Sources/AbsoluteDevicesCapacitor/AbsoluteSecureStoragePlugin.swift
import Capacitor
import Foundation
import Security

@objc(AbsoluteSecureStoragePlugin)
public class AbsoluteSecureStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AbsoluteSecureStoragePlugin"
    public let jsName = "AbsoluteSecureStorage"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "status", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "set", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "keys", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "acquireLease", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseLease", returnType: CAPPluginReturnPromise),
    ]

    private var service: String {
        "\(Bundle.main.bundleIdentifier ?? "absolutejs.app").absolutejs.secure-storage"
    }

    public override func load() {
        let marker = "absolutejs.devices.secure-storage.installed"
        if !UserDefaults.standard.bool(forKey: marker) {
            SecItemDelete([
                kSecClass: kSecClassGenericPassword,
                kSecAttrService: service,
            ] as CFDictionary)
            UserDefaults.standard.set(true, forKey: marker)
        }
    }

    private func requireText(_ call: CAPPluginCall, _ name: String) -> String? {
        guard let value = call.getString(name), !value.isEmpty, value.count <= 256 else {
            call.reject("\(name) must contain between 1 and 256 characters.", "INVALID_ARGUMENT")
            return nil
        }
        return value
    }

    private func query(_ key: String) -> [CFString: Any] {
        [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: key,
            kSecAttrSynchronizable: kCFBooleanFalse as Any,
        ]
    }

    private func rejectStorage(_ call: CAPPluginCall, _ status: OSStatus) {
        call.reject("Native secure storage operation failed (\(status)).", "STORAGE_FAILURE")
    }

    @objc func status(_ call: CAPPluginCall) {
        call.resolve([
            "backend": "keychain",
            "hardwareBacked": false,
            "persistent": true,
            "secure": true,
        ])
    }

    @objc func set(_ call: CAPPluginCall) {
        guard let key = requireText(call, "key") else { return }
        guard let value = call.getString("value") else {
            call.reject("value is required.", "INVALID_ARGUMENT")
            return
        }
        do {
            if let leaseId = call.getString("leaseId") {
                guard try AbsoluteSecureStorageVault.setIfLease(key, value: value, leaseId: leaseId) else {
                    call.reject("Native secure-storage lease was lost.", "LEASE_LOST")
                    return
                }
            } else {
                try AbsoluteSecureStorageVault.set(key, value: value)
            }
            call.resolve()
        } catch {
            call.reject("Native secure storage operation failed.", "STORAGE_FAILURE", error)
        }
    }

    @objc func get(_ call: CAPPluginCall) {
        guard let key = requireText(call, "key") else { return }
        do {
            if let value = try AbsoluteSecureStorageVault.get(key) {
                call.resolve(["value": value])
            } else {
                call.resolve(["value": NSNull()])
            }
        } catch {
            call.reject("Native secure storage operation failed.", "STORAGE_FAILURE", error)
        }
    }

    @objc func remove(_ call: CAPPluginCall) {
        guard let key = requireText(call, "key") else { return }
        do {
            try AbsoluteSecureStorageVault.remove(key)
            call.resolve()
        } catch {
            call.reject("Native secure storage operation failed.", "STORAGE_FAILURE", error)
        }
    }

    private func matchingKeys(prefix: String) -> (keys: [String]?, status: OSStatus) {
        let request: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecReturnAttributes: kCFBooleanTrue as Any,
            kSecMatchLimit: kSecMatchLimitAll,
        ]
        var result: CFTypeRef?
        let status = SecItemCopyMatching(request as CFDictionary, &result)
        if status == errSecItemNotFound { return ([], errSecSuccess) }
        guard status == errSecSuccess else { return (nil, status) }
        let records = result as? [[CFString: Any]] ?? []
        return (
            records.compactMap { $0[kSecAttrAccount] as? String }.filter { $0.hasPrefix(prefix) }.sorted(),
            errSecSuccess
        )
    }

    @objc func keys(_ call: CAPPluginCall) {
        guard let prefix = requireText(call, "prefix") else { return }
        let result = matchingKeys(prefix: prefix)
        if let keys = result.keys {
            call.resolve(["keys": keys])
        } else {
            rejectStorage(call, result.status)
        }
    }

    @objc func clear(_ call: CAPPluginCall) {
        guard let prefix = requireText(call, "prefix") else { return }
        let result = matchingKeys(prefix: prefix)
        guard let keys = result.keys else {
            rejectStorage(call, result.status)
            return
        }
        for key in keys {
            let status = SecItemDelete(query(key) as CFDictionary)
            guard status == errSecSuccess || status == errSecItemNotFound else {
                rejectStorage(call, status)
                return
            }
        }
        call.resolve()
    }

    @objc func acquireLease(_ call: CAPPluginCall) {
        guard let key = requireText(call, "key") else { return }
        let ttl = call.getInt("ttlMs") ?? 0
        guard ttl >= 1_000 && ttl <= 120_000 else {
            call.reject("ttlMs must be between 1000 and 120000.", "INVALID_ARGUMENT")
            return
        }
        if let leaseId = AbsoluteSecureStorageVault.acquireLease(key, ttlMilliseconds: ttl) {
            call.resolve(["leaseId": leaseId])
        } else {
            call.resolve(["leaseId": NSNull()])
        }
    }

    @objc func releaseLease(_ call: CAPPluginCall) {
        guard let key = requireText(call, "key"), let leaseId = requireText(call, "leaseId") else { return }
        AbsoluteSecureStorageVault.releaseLease(key, leaseId: leaseId)
        call.resolve()
    }
}

File: packages/devices-capacitor/android/src/main/java/js/absolute/devices/AbsoluteSecureStoragePlugin.java
package js.absolute.devices;

import android.os.Build;
import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyInfo;
import android.security.keystore.KeyProperties;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.IOException;
import java.security.GeneralSecurityException;
import java.security.KeyStore;
import java.security.spec.InvalidKeySpecException;
import java.util.ArrayList;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;

@CapacitorPlugin(name = "AbsoluteSecureStorage")
public class AbsoluteSecureStoragePlugin extends Plugin {
    private static final String ALIAS = "absolutejs.devices.secure-storage.v1";
    private static final String ANDROID_KEYSTORE = "AndroidKeyStore";
    private static final int MAX_KEY_LENGTH = 256;

    private static String requireText(PluginCall call, String name) {
        String value = call.getString(name);
        if (value == null || value.isEmpty() || value.length() > MAX_KEY_LENGTH) {
            call.reject(name + " must contain between 1 and " + MAX_KEY_LENGTH + " characters.", "INVALID_ARGUMENT");
            return null;
        }
        return value;
    }

    private SecretKey key() throws GeneralSecurityException {
        KeyStore store = KeyStore.getInstance(ANDROID_KEYSTORE);
        try {
            store.load(null);
        } catch (IOException error) {
            throw new GeneralSecurityException("Unable to load Android Keystore.", error);
        }
        SecretKey existing = (SecretKey) store.getKey(ALIAS, null);
        if (existing != null) return existing;

        KeyGenerator generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE);
        generator.init(
            new KeyGenParameterSpec.Builder(
                ALIAS,
                KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setKeySize(256)
                .setRandomizedEncryptionRequired(true)
                .build()
        );
        return generator.generateKey();
    }

    private boolean hardwareBacked() throws GeneralSecurityException {
        try {
            SecretKeyFactory factory = SecretKeyFactory.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE);
            KeyInfo info = (KeyInfo) factory.getKeySpec(key(), KeyInfo.class);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                int level = info.getSecurityLevel();
                return level == KeyProperties.SECURITY_LEVEL_TRUSTED_ENVIRONMENT ||
                    level == KeyProperties.SECURITY_LEVEL_STRONGBOX;
            }
            return info.isInsideSecureHardware();
        } catch (InvalidKeySpecException error) {
            return false;
        }
    }

    private void rejectStorage(PluginCall call, Exception error) {
        call.reject("Native secure storage operation failed.", "STORAGE_FAILURE", error);
    }

    @PluginMethod
    public void status(PluginCall call) {
        try {
            JSObject result = new JSObject();
            result.put("backend", "keystore");
            result.put("hardwareBacked", hardwareBacked());
            result.put("persistent", true);
            result.put("secure", true);
            call.resolve(result);
        } catch (GeneralSecurityException error) {
            rejectStorage(call, error);
        }
    }

    @PluginMethod
    public void set(PluginCall call) {
        String name = requireText(call, "key");
        String value = call.getString("value");
        if (name == null) return;
        if (value == null) {
            call.reject("value is required.", "INVALID_ARGUMENT");
            return;
        }
        try {
            String leaseId = call.getString("leaseId");
            if (leaseId != null && !AbsoluteSecureStorageVault.setIfLease(getContext(), name, value, leaseId)) {
                call.reject("Native secure-storage lease was lost.", "LEASE_LOST");
                return;
            }
            if (leaseId == null) AbsoluteSecureStorageVault.set(getContext(), name, value);
            call.resolve();
        } catch (GeneralSecurityException | IOException error) {
            rejectStorage(call, error);
        }
    }

    @PluginMethod
    public void get(PluginCall call) {
        String name = requireText(call, "key");
        if (name == null) return;
        try {
            JSObject result = new JSObject();
            String value = AbsoluteSecureStorageVault.get(getContext(), name);
            result.put("value", value == null ? JSObject.NULL : value);
            call.resolve(result);
        } catch (GeneralSecurityException | IOException error) {
            rejectStorage(call, error);
        }
    }

    @PluginMethod
    public void remove(PluginCall call) {
        String name = requireText(call, "key");
        if (name == null) return;
        try {
            AbsoluteSecureStorageVault.remove(getContext(), name);
            call.resolve();
        } catch (IOException error) {
            rejectStorage(call, error);
        }
    }

    @PluginMethod
    public void keys(PluginCall call) {
        String prefix = requireText(call, "prefix");
        if (prefix == null) return;
        try {
            ArrayList<String> matches = AbsoluteSecureStorageVault.keys(getContext(), prefix);
            JSObject result = new JSObject();
            result.put("keys", new JSArray(matches));
            call.resolve(result);
        } catch (IOException error) {
            rejectStorage(call, error);
        }
    }

    @PluginMethod
    public void clear(PluginCall call) {
        String prefix = requireText(call, "prefix");
        if (prefix == null) return;
        try {
            AbsoluteSecureStorageVault.clear(getContext(), prefix);
            call.resolve();
        } catch (IOException error) {
            rejectStorage(call, error);
        }
    }

    @PluginMethod
    public void acquireLease(PluginCall call) {
        String name = requireText(call, "key");
        Integer ttl = call.getInt("ttlMs");
        if (name == null) return;
        if (ttl == null || ttl < 1000 || ttl > 120000) {
            call.reject("ttlMs must be between 1000 and 120000.", "INVALID_ARGUMENT");
            return;
        }
        JSObject result = new JSObject();
        String leaseId = AbsoluteSecureStorageVault.acquireLease(name, ttl);
        result.put("leaseId", leaseId == null ? JSObject.NULL : leaseId);
        call.resolve(result);
    }

    @PluginMethod
    public void releaseLease(PluginCall call) {
        String name = requireText(call, "key");
        String leaseId = requireText(call, "leaseId");
        if (name == null || leaseId == null) return;
        AbsoluteSecureStorageVault.releaseLease(name, leaseId);
        call.resolve();
    }
}
